pyomo/core/base/componentuid.py

```diff
@@ -0,0 +1,576 @@
+                else:
+                    assert tok.type in {'WORD','STRING','NUMBER','PICKLE'}
+                    idx_stack[-1].append(tok.value)
+            else:
+                assert tok.type in {'WORD','STRING'}
+                assert name is None
+                name = tok.value
+        assert not idx_stack
+        yield (name, idx)
+            
+    def _parse_cuid_v1(self, label):
+        """Parse a string (v1 repr format) and yield name, idx pairs
+
+        This attempts to parse a string (nominally returned by
+        get_repr()) to generate the sequence of (name, idx) pairs for
+        the _cuids data structure.
+
+        """
+        cList = label.split('.')
+        for c in cList:
+            if c[-1] == ']':
+                c_info = c[:-1].split('[',1)
+            else:
+                c_info = c.split(':',1)
+            if len(c_info) == 1:
+                yield ( c_info[0], tuple() )
+            else:
+                idx = c_info[1].split(',')
+                for i, val in enumerate(idx):
+                    if val == '*':
+                        idx[i] = slice(None)
+                    elif val[0] == '$':
+                        idx[i] = str(val[1:])
+                    elif val[0] == '#':
+                        idx[i] = _int_or_float(val[1:])
+                    elif val[0] in  "\"'" and val[-1] == val[0]:
+                        idx[i] = val[1:-1]
+                    elif _re_number.match(val):
+                        idx[i] = _int_or_float(val)
+                if len(idx) == 1 and idx[0] == '**':
+                    yield ( c_info[0], (Ellipsis,) )
+                else:
+                    yield ( c_info[0], tuple(idx) )
+
+    def _resolve_cuid(self, block):
+        obj = block
+        for name, idx in self._cids:
+            try:
+                if not idx:
+                    obj = getattr(obj, name)
+                elif len(idx) == 1:
+                    obj = getattr(obj, name)[idx[0]]
+                else:
+                    obj = getattr(obj, name)[idx]
+            except KeyError:
+                return None
+            except AttributeError:
+                return None
+            except IndexError:
+                return None
+        return obj
+
+    @deprecated("ComponentUID.find_component() is deprecated. "
+                "Use ComponentUID.find_component_on()", version='TBD')
+    def find_component(self, block):
+        return self.find_component_on(block)
+
+    def find_component_on(self, block):
+        """
+        Return the (unique) component in the block.  If the CUID contains
+        a wildcard in the last component, then returns that component.  If
+        there are wildcards elsewhere (or the last component was a partial
+        slice), then returns None.  See list_components below.
+        """
+        obj = self._resolve_cuid(block)
+        if isinstance(obj, IndexedComponent_slice):
+            return None
```

Comment:
I had thought about the same thing and originally elected to leave the behavior unchanged.  However,I think that is sufficiently useful that the API change is a good idea.

This turned out to be a little challenging to implement because it introduced a circular import for Python < 3.7.  I figured out how to make a more robust `relocated_module_attribute()` method that breaks the circular import for all versions of Python.